File: protocol/src/test/csharp/zfoocs/ProtocolManager.cs
using System;
using System.Collections.Generic;

namespace zfoocs
{
    public class ProtocolManager
    {
        public static readonly short MAX_PROTOCOL_NUM = short.MaxValue;


        private static readonly IProtocolRegistration[] protocols = new IProtocolRegistration[MAX_PROTOCOL_NUM];
        private static readonly Dictionary<Type, short> protocolIdMap = new Dictionary<Type, short>();
        private static bool initialized = false;


        public static void InitProtocol()
        {
            if (initialized)
            {
                return;
            }
            initialized = true;
            protocols[0] = new EmptyObjectRegistration();
            protocolIdMap[typeof(EmptyObject)] = 0;
            protocols[1] = new VeryBigObjectRegistration();
            protocolIdMap[typeof(VeryBigObject)] = 1;
            protocols[100] = new ComplexObjectRegistration();
            protocolIdMap[typeof(ComplexObject)] = 100;
            protocols[101] = new NormalObjectRegistration();
            protocolIdMap[typeof(NormalObject)] = 101;
            protocols[102] = new ObjectARegistration();
            protocolIdMap[typeof(ObjectA)] = 102;
            protocols[103] = new ObjectBRegistration();
            protocolIdMap[typeof(ObjectB)] = 103;
            protocols[104] = new SimpleObjectRegistration();
            protocolIdMap[typeof(SimpleObject)] = 104;
        }

        public static short GetProtocolId(Type type)
        {
            return protocolIdMap[type];
        }

        public static IProtocolRegistration GetProtocol(short protocolId)
        {
            var protocol = protocols[protocolId];
            if (protocol == null)
            {
                throw new Exception("[protocolId:" + protocolId + "] not exist");
            }
            return protocol;
        }

        public static void Write(ByteBuffer buffer, object packet)
        {
            var protocolId = protocolIdMap[packet.GetType()];
            // write protocol id to buffer
            buffer.WriteShort(protocolId);
            // write packet
            GetProtocol(protocolId).Write(buffer, packet);
        }

        public static object Read(ByteBuffer buffer)
        {
            var protocolId = buffer.ReadShort();
            return GetProtocol(protocolId).Read(buffer);
        }
    }
}

File: protocol/src/test/kotlin/com/zfoo/kotlin/ProtocolManager.kt
package com.zfoo.kotlin
import com.zfoo.kotlin.packet.EmptyObject
import com.zfoo.kotlin.packet.EmptyObjectRegistration
import com.zfoo.kotlin.packet.ComplexObject
import com.zfoo.kotlin.packet.ComplexObjectRegistration
import com.zfoo.kotlin.packet.NormalObject
import com.zfoo.kotlin.packet.NormalObjectRegistration
import com.zfoo.kotlin.packet.ObjectA
import com.zfoo.kotlin.packet.ObjectARegistration
import com.zfoo.kotlin.packet.ObjectB
import com.zfoo.kotlin.packet.ObjectBRegistration
import com.zfoo.kotlin.packet.SimpleObject
import com.zfoo.kotlin.packet.SimpleObjectRegistration
class ProtocolManager {
    companion object {
        val protocols = arrayOfNulls<IProtocolRegistration>(Short.MAX_VALUE.toInt())
        var protocolIdMap: MutableMap<Class<*>, Short> = HashMap()
        @JvmStatic
        fun initProtocol() {
            // initProtocol
            protocols[0] = EmptyObjectRegistration()
            protocolIdMap[EmptyObject::class.java] = 0.toShort()
            protocols[100] = ComplexObjectRegistration()
            protocolIdMap[ComplexObject::class.java] = 100.toShort()
            protocols[101] = NormalObjectRegistration()
            protocolIdMap[NormalObject::class.java] = 101.toShort()
            protocols[102] = ObjectARegistration()
            protocolIdMap[ObjectA::class.java] = 102.toShort()
            protocols[103] = ObjectBRegistration()
            protocolIdMap[ObjectB::class.java] = 103.toShort()
            protocols[104] = SimpleObjectRegistration()
            protocolIdMap[SimpleObject::class.java] = 104.toShort()
        }

        @JvmStatic
        fun getProtocolId(clazz: Class<*>): Short {
            return protocolIdMap[clazz]!!
        }

        @JvmStatic
        fun getProtocol(protocolId: Short): IProtocolRegistration {
            return protocols[protocolId.toInt()]
                ?: throw RuntimeException("[protocolId:$protocolId] not exist")
        }

        @JvmStatic
        fun write(buffer: ByteBuffer, packet: Any) {
            val protocolId = getProtocolId(packet.javaClass)
            // write protocol id to buffer
            buffer.writeShort(protocolId)
            // write packet
            getProtocol(protocolId).write(buffer, packet)
        }

        @JvmStatic
        fun read(buffer: ByteBuffer): Any {
            val protocolId = buffer.readShort()
            return getProtocol(protocolId).read(buffer)
        }
    }
}